### PR TITLE
Cache get request with wildcard

### PR DESCRIPTION
**Pull request by harshkothari410**
- Added functionality to Cache GET request of samples with wildcard
- Added Test
- Added Flag to turn it on this feature
- Added invalidate timeout flag: default is 5 second

### CODE REVIEW

File: api/v1/controllers/samples.js [hunk 1]
```diff
@@ -57,21 +93,33 @@ module.exports = {
...
-        }
-
-        u.logAPI(req, resultObj, response); // audit log
-        res.status(httpStatus.OK).json(response);
-      })
-      .catch((err) => u.handleError(next, err, helper.modelName));
+        });
+      } else {
+        doFindSampleStoreResponse(req, res,
+          next, resultObj);
```
Comment (iamigo) [inline api/v1/controllers/samples.js]: no line break before `next, resultObj);`

File: api/v1/controllers/samples.js [hunk 2]
```diff
@@ -57,21 +93,33 @@ module.exports = {
    * @param {Function} next - The next middleware function in the stack
    */
   findSamples(req, res, next) {
+    // Check if Cache is on for Wildcard Sample query
+    if (featureToggles.isFeatureEnabled('cacheGetSamplesWildcard')) {
+      const query = req.query.name;
+      helper.cacheEnabled =
+        query ? (query.indexOf('*') > -1) ? true : false : false;
```
Comment (iamigo) [inline api/v1/controllers/samples.js]: I had to read this line multiple times to understand it. Please use extra parentheses to make it more readable, or break into two statements.

File: config.js [hunk 3]
```diff
@@ -50,6 +50,9 @@ const iplist = configUtil.parseIPlist(ipWhitelist);
 // Check for timed-out samples every 30 seconds if not specified in env var
 const DEFAULT_CHECK_TIMEOUT_INTERVAL_MILLIS = 30000;
 
+// GET Samples Cache invalidation default time
+const GET_SAMPLES_WILDCARD_CACHE_INVALIDATION = 5;
```
Comment (iamigo) [inline config.js]: call this DEFAULT_GET_SAMPLES_WILDCARD_CACHE_INVALIDATION

File: api/v1/controllers/samples.js [hunk 4]
```diff
@@ -57,21 +93,33 @@ module.exports = {
    * @param {Function} next - The next middleware function in the stack
    */
   findSamples(req, res, next) {
+    // Check if Cache is on for Wildcard Sample query
+    if (featureToggles.isFeatureEnabled('cacheGetSamplesWildcard')) {
+      const query = req.query.name;
```
Comment (iamigo) [inline api/v1/controllers/samples.js]: so we are only using this cache for wildcard queries on the "name" field? if so, please consider renaming the feature toggle to "cacheGetSamplesByNameWildcard" and updating any relevant comments

Comment (harshkothari410) [reply api/v1/controllers/samples.js]: Yes we decided only on name.

File: api/v1/controllers/samples.js (hunk 2 above)
Comment (harshkothari410) [reply api/v1/controllers/samples.js]: Got it 👍

File: config.js (hunk 3 above)
Comment (harshkothari410) [reply config.js]: ok

File: api/v1/controllers/samples.js (hunk 4 above)
Comment (shriramshankar) [inline api/v1/controllers/samples.js]: if think if you do something like `const query = req.query.name || ''`; you can avoid the nested ternary in line 100.

File: api/v1/controllers/samples.js [hunk 5]
```diff
@@ -57,21 +93,33 @@ module.exports = {
...
+      helper.cacheEnabled =
+        query ? (query.indexOf('*') > -1) ? true : false : false;
+      helper.cacheKey =  helper.cacheEnabled ? query : null;
+      helper.cacheExpiry = helper.cacheEnabled ?
+        parseInt(getSamplesWildcardCacheInvalidation) : null;
```
Comment (shriramshankar) [inline api/v1/controllers/samples.js]: Is specifying the radix required in parseInt?

File: api/v1/controllers/samples.js (hunk 4 above)
Comment (shriramshankar) [reply api/v1/controllers/samples.js]: or changing it to 
`helper.cacheEnabled = query && (query.indexOf('*') > -1)` , will give the same effect. Correct?

File: tests/cache/models/samples/get.js [hunk 6]
```diff
@@ -76,6 +77,73 @@ describe('tests/cache/models/samples/get.js, ' +
...
+
+        expect(JSON.parse(reply).length).to.be.equal(3);
+        expect(JSON.parse(reply)[0].name).to.equal(s1s2a1);
+        done();
+      });
+    });
+  });
+
+  it('get without wildcard and with cacheGetSamplesWildcard flag on ' +
+    'shold not cache response', (done) => {
```
Comment (shriramshankar) [inline tests/cache/models/samples/get.js]: "should". Same with the other tests above.

File: tests/cache/models/samples/get.js [hunk 7]
```diff
@@ -76,6 +77,73 @@ describe('tests/cache/models/samples/get.js, ' +
...
+    tu.toggleOverride('cacheGetSamplesWildcard', false);
+    api.get(`${path}?name=___Subj*`)
+    .set('Authorization', token)
+    .expect(constants.httpStatus.OK)
+    .end((err, res) => {
+      if (err) {
+        return done(err);
+      }
+
+      redisCache.get('___Subj*', (cacheErr, reply) => {
```
Comment (shriramshankar) [inline tests/cache/models/samples/get.js]: Any reason behind accessing Redis directly here, instead of using the api?

Comment (harshkothari410) [reply tests/cache/models/samples/get.js]: Yes so that I can verify that it is cached.

File: api/v1/controllers/samples.js (hunk 4 above)
Comment (harshkothari410) [reply api/v1/controllers/samples.js]: Good point. Correct. Thanks :)

File: api/v1/controllers/samples.js (hunk 5 above)
Comment (harshkothari410) [reply api/v1/controllers/samples.js]: Yes time in second.

File: api/v1/controllers/samples.js [hunk 8]
```diff
@@ -94,10 +94,9 @@ module.exports = {
    */
   findSamples(req, res, next) {
     // Check if Cache is on for Wildcard Sample query
-    if (featureToggles.isFeatureEnabled('cacheGetSamplesWildcard')) {
-      const query = req.query.name;
-      helper.cacheEnabled =
-        query ? (query.indexOf('*') > -1) ? true : false : false;
+    if (featureToggles.isFeatureEnabled('cacheGetSamplesByNameWildcard')) {
+      const query = req.query.name || '';
```
Comment (shriramshankar) [inline api/v1/controllers/samples.js]: I think I was not clear in my comments. You do not need to do `|| ''`., just the change you have on line 99 is enough.

File: tests/cache/models/samples/get.js (hunk 6 above)
Comment (shriramshankar) [reply tests/cache/models/samples/get.js]: @harshkothari410 Looks like you missed this comment.

**Comment by shriramshankar** [comment]
Ok to merge, once the tests pass.

**Comment by iamigo** [comment]
@harshkothari410 build failing?

**Comment by iamigo** [comment]
@harshkothari410 ?

File: tests/api/v1/samples/get.js [hunk 9]
```diff
@@ -83,6 +84,70 @@ describe(`tests/api/v1/samples/get.js, GET ${path} >`, () => {
     .end(done);
   });
 
+  it('get with wildcard and with cacheGetSamplesWildcard flag on ' +
+    'shold cache response', (done) => {
+    tu.toggleOverride('cacheGetSamplesByNameWildcard', true);
```
[the remark this answers is not in the document]
Comment (iamigo) [reply tests/api/v1/samples/get.js]: >>>> toggle override here without storing original value and resetting it back to that value

File: tests/cache/models/samples/get.js [hunk 10]
```diff
@@ -76,6 +80,80 @@ describe('tests/cache/models/samples/get.js, ' +
     });
   });
 
+  it('get with wildcard and with cacheGetSamplesWildcard flag on ' +
+    'should cache response', (done) => {
+    tu.toggleOverride('cacheGetSamplesByNameWildcard', true);
```
Comment (iamigo) [inline tests/cache/models/samples/get.js]: toggle override here without storing original value and resetting it back to that value

File: tests/enableCache/perspectives.js [hunk 11]
```diff
@@ -76,17 +78,54 @@ describe(`tests/enableCache/perspectives.js, api: GET ${path} >`, () => {
...
+          console.log(cacheErr);
+          throw new Error('Expected response value in cache');
+        }
+      });
+    });
+  });
+
+  it('basic get, response present in cache with enableApiActivityLogs',
+    (done) => {
+    tu.toggleOverride('enableApiActivityLogs', true);
```
Comment (iamigo) [inline tests/enableCache/perspectives.js]: toggleOverride here without saving previous feature enablement value and restoring it after

File: tests/api/v1/samples/get.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (harshkothari410) [reply tests/api/v1/samples/get.js]: see the line number 24 25 @iamigo

[the remark this answers is not in the document]
Comment (iamigo) [reply tests/api/v1/samples/get.js]: Line 24 is the "describe" -- what should I be looking for?

File: api/v1/controllers/perspectives.js [hunk 12]
```diff
@@ -79,8 +81,15 @@ module.exports = {
    * @param {Function} next - The next middleware function in the stack
    */
   findPerspectives(req, res, next) {
-    helper.cacheEnabled = featureToggles
-    .isFeatureEnabled('enableCachePerspective');
+    // Caching perspective
+    if (featureToggles
+      .isFeatureEnabled('enableCachePerspective') &&
```
Comment (iamigo) [inline api/v1/controllers/perspectives.js]: line 84 and 85 should be on the same line

File: api/v1/controllers/samples.js [hunk 13]
```diff
@@ -31,6 +31,42 @@ const utils = require('./utils');
 const publisher = u.publisher;
 const kueSetup = require('../../../jobQueue/setup');
 const kue = kueSetup.kue;
+const getSamplesWildcardCacheInvalidation = require('../../../config')
+  .getSamplesWildcardCacheInvalidation;
+const redisCache = require('../../../cache/redisCache').client.cache;
+
+/**
+ * Function for finding sample from samplestore. If cache is on then
```
Comment (iamigo) [inline api/v1/controllers/samples.js]: No need to say "function for..." in the function-level comment. Should just be `Find sample from samplestore. If cache is on then cache the response as well'

File: api/v1/controllers/samples.js [hunk 14]
```diff
@@ -31,6 +31,42 @@ const utils = require('./utils');
...
+ * @param {IncomingMessage} req - The request object
+ * @param {ServerResponse} res - The response object
+ * @param {Function} next - The next middleware function in the stack
+ * @param {Object} resultObj - DateTime object for logging
+ * @param {String} cacheKey - Cache Key
+ * @param {Integer} cacheExpiry -  Cache expiry time
+ */
+function doFindSampleStoreResponse(req, res, next, resultObj, cacheKey, cacheExpiry) {
+  redisModelSample.findSamples(req, res, resultObj)
+    .then((response) => {
```
Comment (iamigo) [inline api/v1/controllers/samples.js]: lines 51-68 should have same indent as line 50

File: api/v1/helpers/verbs/doFind.js [hunk 15]
```diff
@@ -71,7 +70,7 @@ function doFindAndCountAll(reqResNext, props, opts) {
  * @param {Object} cacheKey - Optional cache key used to cache the response in
  * redis
  */
-function doFindResponse(reqResNext, props, opts, cacheKey) {
+function doFindResponse(reqResNext, props, opts, cacheKey, cacheExpiry) {
```
Comment (iamigo) [inline api/v1/helpers/verbs/doFind.js]: add the new param to the list of params in the function comments

File: tests/api/v1/samples/get.js [hunk 16]
```diff
@@ -83,6 +87,79 @@ describe(`tests/api/v1/samples/get.js, GET ${path} >`, () => {
...
+      redisCache.get(`${sampleName}*`, (cacheErr, reply) => {
+        if (cacheErr || !reply) {
+          return done(cacheErr);
+        }
+
+        expect(JSON.parse(reply).length).to.be.above(ZERO);
+        expect(JSON.parse(reply)[0].name).to.equal(`${sampleName}`);
+        redisCache.del(`${sampleName}*`);
+
+        tu.toggleOverride('cacheGetSamplesByNameWildcard',
```
Comment (iamigo) [inline tests/api/v1/samples/get.js]: Putting it here means the value doesn't get restored to its original value if we returned done from line 98 or line 103. Saving the previous value of a feature toggle and restoring that original value should go in before/after blocks.

File: tests/api/v1/samples/get.js [hunk 17]
```diff
@@ -83,6 +87,79 @@ describe(`tests/api/v1/samples/get.js, GET ${path} >`, () => {
...
+    .end((err, res) => {
+      if (err) {
+        return done(err);
+      }
+
+      redisCache.get(`${sampleName}`, (cacheErr, reply) => {
+        if (cacheErr || !reply) {
+          expect(res.body.length).to.be.above(ZERO);
+
+          tu.toggleOverride('cacheGetSamplesByNameWildcard',
```
Comment (iamigo) [inline tests/api/v1/samples/get.js]: same as other comment

File: tests/api/v1/samples/get.js [hunk 18]
```diff
@@ -83,6 +87,79 @@ describe(`tests/api/v1/samples/get.js, GET ${path} >`, () => {
...
+    .end((err, res) => {
+      if (err) {
+        return done(err);
+      }
+
+      redisCache.get(`${sampleName}*`, (cacheErr, reply) => {
+        if (cacheErr || !reply) {
+          expect(res.body.length).to.be.above(ZERO);
+
+          tu.toggleOverride('cacheGetSamplesByNameWildcard',
```
Comment (iamigo) [inline tests/api/v1/samples/get.js]: same as other comment

File: tests/cache/models/samples/get.js [hunk 19]
```diff
@@ -76,6 +80,80 @@ describe('tests/cache/models/samples/get.js, ' +
...
+
+      redisCache.get('___Subj*', (cacheErr, reply) => {
+        if (cacheErr || !reply) {
+          return done(cacheErr);
+        }
+
+        expect(JSON.parse(reply).length).to.be.equal(3);
+        expect(JSON.parse(reply)[0].name).to.equal(s1s2a1);
+        redisCache.del('___Subj*');
+        tu.toggleOverride('cacheGetSamplesByNameWildcard',
```
Comment (iamigo) [inline tests/cache/models/samples/get.js]: same as other comment

File: tests/cache/models/samples/get.js [hunk 20]
```diff
@@ -76,6 +80,80 @@ describe('tests/cache/models/samples/get.js, ' +
...
+      if (err) {
+        return done(err);
+      }
+
+      redisCache.get('___Subject1.___Subject2|___Aspect1',
+        (cacheErr, reply) => {
+        if (cacheErr || !reply) {
+          expect(res.body.length).to.be.equal(1);
+          expect(res.body[0].name).to.equal(s1s2a1);
+          tu.toggleOverride('cacheGetSamplesByNameWildcard',
```
Comment (iamigo) [inline tests/cache/models/samples/get.js]: same as other comment

File: tests/cache/models/samples/get.js [hunk 21]
```diff
@@ -76,6 +80,80 @@ describe('tests/cache/models/samples/get.js, ' +
...
+    .end((err, res) => {
+      if (err) {
+        return done(err);
+      }
+
+      redisCache.get('___Subj*', (cacheErr, reply) => {
+        if (cacheErr || !reply) {
+          expect(res.body.length).to.be.equal(3);
+          expect(res.body[0].name).to.equal(s1s2a1);
+          tu.toggleOverride('cacheGetSamplesByNameWildcard',
```
Comment (iamigo) [inline tests/cache/models/samples/get.js]: same as other comment

File: tests/enableCache/perspectives.js [hunk 22]
```diff
@@ -76,17 +79,54 @@ describe(`tests/enableCache/perspectives.js, api: GET ${path} >`, () => {
...
-        const jsonReply = JSON.parse(reply);
-        expect(jsonReply).to.have.length(ONE);
-        expect(jsonReply[ZERO].name).to.be.equal('___testPersp');
-        expect(jsonReply).to.have.deep.property('[0].lensId', lensId);
+    api.get(path)
+    .set('Authorization', token)
+    .expect(constants.httpStatus.OK)
+    .end((err, res) => {
+      if (err) {
+        done(err);
```
Comment (iamigo) [inline tests/enableCache/perspectives.js]: return done (err);

File: tests/enableCache/perspectives.js [hunk 23]
```diff
@@ -76,17 +79,54 @@ describe(`tests/enableCache/perspectives.js, api: GET ${path} >`, () => {
...
-        done();
-      } else {
-        throw new Error('Expected response value in cache');
+      redisCache.get('/v1/perspectives', (cacheErr, reply) => {
+        if (reply) {
+          const jsonReply = JSON.parse(reply);
+          expect(jsonReply).to.have.length(ONE);
+          expect(jsonReply[ZERO].name).to.be.equal('___testPersp');
+          expect(jsonReply).to.have.deep.property('[0].lensId', lensId);
+
```
Comment (iamigo) [inline tests/enableCache/perspectives.js]: no line break here

File: tests/enableCache/perspectives.js [hunk 24]
```diff
@@ -76,17 +79,54 @@ describe(`tests/enableCache/perspectives.js, api: GET ${path} >`, () => {
...
+          done();
```
Comment (iamigo) [inline tests/enableCache/perspectives.js]: return done();

File: tests/enableCache/perspectives.js [hunk 25]
```diff
@@ -76,17 +79,54 @@ describe(`tests/enableCache/perspectives.js, api: GET ${path} >`, () => {
...
+        } else {
+          console.log(cacheErr);
```
Comment (iamigo) [inline tests/enableCache/perspectives.js]: no console log

File: tests/enableCache/perspectives.js [hunk 26]
```diff
@@ -76,17 +79,54 @@ describe(`tests/enableCache/perspectives.js, api: GET ${path} >`, () => {
...
+        done(err);
       }
+
+      redisCache.get('/v1/perspectives', (cacheErr, reply) => {
+        if (reply) {
+          const jsonReply = JSON.parse(reply);
+          expect(jsonReply).to.have.length(ONE);
+          expect(jsonReply[ZERO].name).to.be.equal('___testPersp');
+          expect(jsonReply).to.have.deep.property('[0].lensId', lensId);
+          tu.toggleOverride('enableApiActivityLogs', enableApiActivityLogs);
```
Comment (iamigo) [inline tests/enableCache/perspectives.js]: same as other comment

File: tests/enableCache/perspectives.js [hunk 27]
```diff
@@ -76,17 +79,54 @@ describe(`tests/enableCache/perspectives.js, api: GET ${path} >`, () => {
...
+          done();
+        } else {
+          console.log(cacheErr);
```
Comment (iamigo) [inline tests/enableCache/perspectives.js]: no console log

File: tests/enableCache/perspectives.js [hunk 28]
```diff
@@ -108,11 +148,20 @@ describe(`tests/enableCache/perspectives.js, api: GET ${path} >`, () => {
   });
 
   it('get with limit and sort, response not present in cache', (done) => {
-    redisCache.get('{"order":["name"],"limit":10,"where":{}}',
+    api.get(path + '?limit=10&sort=name')
+    .set('Authorization', token)
+    .expect(constants.httpStatus.OK)
+    .end((err, res) => {
+      if (err) {
+        done(err);
```
Comment (iamigo) [inline tests/enableCache/perspectives.js]: `return done(err);`

File: tests/api/v1/samples/get.js (hunk 16 above)
Comment (harshkothari410) [reply tests/api/v1/samples/get.js]: But I don't need in each and every test. So there is no point of it writing in before / after.

Comment (iamigo) [reply tests/api/v1/samples/get.js]: The only way to make sure they get called no matter the outcome of the test is to use before/after. Put the ones which need it in their own nested "describe" and add the before/after in that describe only.

File: tests/cache/models/samples/get.js [hunk 29]
```diff
@@ -582,3 +511,123 @@ describe('tests/cache/models/samples/get.js, ' +
...
+    .set('Authorization', token)
+    .expect(constants.httpStatus.OK)
+    .end((err, res) => {
+      if (err) {
+        return done(err);
+      }
+
+      redisCache.get('___Subj*', (cacheErr, reply) => {
+        if (cacheErr || !reply) {
+          return done(cacheErr);
```
Comment (iamigo) [inline tests/cache/models/samples/get.js]: If cacheErr is false-y here, it means we have `!reply`, but we're returning `done(cacheErr)` (with an empty `cacheErr`, and calling done with nothing is treated as test success, not test failure.

File: tests/api/v1/samples/get.js [hunk 30]
```diff
@@ -150,74 +193,84 @@ describe(`tests/api/v1/samples/get.js, GET ${path} >`, () => {
...
+        expect(JSON.parse(reply)[0].name).to.equal(`${sampleName}`);
+        redisCache.del(`${sampleName}*`);
+        done();
       });
     });
   });
 
-  it('basic get does not return id', (done) => {
-    api.get(path)
+  it('get without wildcard shold not cache response', (done) => {
```
Comment (iamigo) [inline tests/api/v1/samples/get.js]: "should"

File: tests/api/v1/samples/get.js [hunk 31]
```diff
@@ -87,59 +87,102 @@ describe(`tests/api/v1/samples/get.js, GET ${path} >`, () => {
...
+      }
+    })
+    .end(done);
   });
 
-  it('get without wildcard and with cacheGetSamplesWildcard flag on ' +
-    'shold not cache response', (done) => {
-    tu.toggleOverride('cacheGetSamplesByNameWildcard', true);
-    api.get(`${path}?name=${sampleName}`)
+  it('by name is case in-sensitive', (done) => {
```
Comment (iamigo) [inline tests/api/v1/samples/get.js]: "insensitive"

File: tests/enableCache/perspectives.js [hunk 32]
```diff
@@ -188,8 +183,7 @@ describe(`tests/enableCache/perspectives.js, api: GET ${path} >`, () => {
           expect(jsonReply.name).to.equal(`${tu.namePrefix}testPersp`);
           expect(jsonReply.rootSubject).to.equal('myMainSubject');
           expect(jsonReply.lensId).to.equal(lensId);
-
-          done();
+          return done();
         } else {
           throw new Error('Expected response value in cache');
```
Comment (iamigo) [inline tests/enableCache/perspectives.js]: our standard test semantics is to return `done(...)` with some value, e.g. `done('Expected response value in cache');`

**Pull request merged**